### PR TITLE
Fix/snap sync edge cases

### PR DESCRIPTION
**Pull request by asdacap**
- Allow increasing partition count above 256, up to int.MaxValue. This is mainly useful for triggering edge case.
- Collect, and add edge case to unit test and fix them.
- Fix storage get requested twice between partition boundary.
- Fix storage may get persisted twice between partition boundary.
- Fix the same two above issue between address range (the first and last key always overlap). 
- Switch boundary logic to use `TreePath` instead of array.

#### What types of changes does your code introduce?

- [X] Bugfix (a non-breaking change that fixes an issue)

#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [X] Yes
- [ ] No

#### Notes on testing

- Testing...

### CODE REVIEW

File: src/Nethermind/Nethermind.Synchronization/SnapSync/ProgressTracker.cs [hunk 1]
```diff
@@ -72,8 +73,8 @@ public ProgressTracker(IBlockTree blockTree, IDb db, ILogManager logManager, int
 
             _pivot = new Pivot(blockTree, logManager);
 
-            if (accountRangePartitionCount < 1 || accountRangePartitionCount > 256)
-                throw new ArgumentException("Account range partition must be between 1 to 256.");
+            if (accountRangePartitionCount < 1 || accountRangePartitionCount > int.MaxValue)
+                throw new ArgumentException($"Account range partition must be between 1 to {int.MaxValue}.");
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Synchronization/SnapSync/ProgressTracker.cs:77]: True, its useless now.

File: src/Nethermind/Nethermind.Synchronization/SnapSync/SnapProviderHelper.cs [hunk 2]
```diff
@@ -72,7 +80,30 @@ public static (AddRangeResult result, bool moreChildrenToRight, List<PathWithAcc
...
+                // Fortunately, this should only happen n-1 time where n is the number of top level
+                // partition count.
+
+                tree.RootHash = Keccak.EmptyTreeHash;
+                for (var index = 0; index < accounts.Count; index++)
+                {
+                    PathWithAccount account = accounts[index];
+                    if (account.Path >= limitHash || account.Path < startingHash) continue;
+                    _ = tree.Set(account.Path, account.Account);
+                }
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Synchronization/SnapSync/SnapProviderHelper.cs:101]: Its not just the node, but the parents that touches the node also. Partitionning is not useless, this only happens 16 time on mainnet.

[the remark this answers is not in the document]
Comment (damian-orzechowski) [reply src/Nethermind/Nethermind.Synchronization/SnapSync/SnapProviderHelper.cs:101]: Probably a super minor optimisation, but could check `hasExtraStorage` in the 1st loop and not make changes to the tree, if we will still will rebuild it.

[the remark this answers is not in the document]
Comment (damian-orzechowski) [reply src/Nethermind/Nethermind.Synchronization/SnapSync/SnapProviderHelper.cs:101]: Also - if we skip the extra node, why then skip stitching as well?

File: src/Nethermind/Nethermind.Core/Crypto/Hash256.cs [hunk 3]
```diff
@@ -58,6 +59,10 @@ public ValueHash256(ReadOnlySpan<byte> bytes)
             _bytes = Unsafe.As<byte, Vector256<byte>>(ref MemoryMarshal.GetReference(bytes));
         }
 
+        public ValueHash256(UInt256 uint256) : this(uint256.ToBigEndian())
```
Comment (tanishqjasoria) [inline src/Nethermind/Nethermind.Core/Crypto/Hash256.cs:62]: i think we should add `bool isBigEndian = true` parameter here

File: src/Nethermind/Nethermind.Core/Crypto/Hash256.cs [hunk 4]
```diff
@@ -112,6 +117,11 @@ public Hash256 ToCommitment()
         public static bool operator >=(in ValueHash256 left, in ValueHash256 right) => left.CompareTo(in right) >= 0;
         public static bool operator <=(in ValueHash256 left, in ValueHash256 right) => left.CompareTo(in right) <= 0;
         public static implicit operator Hash256(in ValueHash256 keccak) => new(keccak);
+
+        public UInt256 ToUInt256()
+        {
+            return new UInt256(Bytes, isBigEndian: true);
```
Comment (tanishqjasoria) [inline src/Nethermind/Nethermind.Core/Crypto/Hash256.cs:123]: same here, `bool isBigEndian = true`

**Review by tanishqjasoria — APPROVED** [review]

**Review by benaadams — APPROVED** [review]

File: src/Nethermind/Nethermind.Core/Crypto/Hash256.cs (hunk 3 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Core/Crypto/Hash256.cs]: done
